packages/core/src/crypto/x509/X509Certificate.ts

```diff
@@ -0,0 +1,134 @@
+  }
+
+  public static fromEncodedCertificate(encodedCertificate: string): X509Certificate {
+    const certificate = new x509.X509Certificate(encodedCertificate)
+    return this.parseCertificate(certificate)
+  }
+
+  private static parseCertificate(certificate: x509.X509Certificate): X509Certificate {
+    // TODO: this breaks as for P-256 it will be `{name: 'ECDSA', namedCurve: 'P-256'}`
+    // but for ed25519 it will be `[name: 'ed25519']`
+    const algorithm = certificate.publicKey.algorithm.namedCurve
+      ? certificate.publicKey.algorithm.namedCurve
+      : certificate.publicKey.algorithm.name
+
+    const publicKey = AsnParser.parse(certificate.publicKey.rawData, SubjectPublicKeyInfo)
+    const privateKey = certificate.privateKey ? new Uint8Array(certificate.privateKey.rawData) : undefined
+
+    // TODO: remove this and validate properly from webcrypto-core algorithm to keyType of credo
+    if (algorithm.toLowerCase() !== 'ecdsa') throw new X509Error(`Expected ECDSA algorithm, received: ${algorithm}`)
+
+    const key = new Key(new Uint8Array(publicKey.subjectPublicKey), KeyType.P256)
+
+    return new X509Certificate({
+      publicKey: key,
+      privateKey,
+      extensions: certificate.extensions
+        ?.map((e) => JSON.parse(JSON.stringify(e)))
+        .map((e) => ({ [e.type]: e.names })) as Array<Extension>,
+      rawCertificate: new Uint8Array(certificate.rawData),
+    })
+  }
+
+  private getMatchingExtensions<T>(name: string, type: string): Array<T> | undefined {
+    const extensionsWithName = this.extensions
+      ?.filter((e) => e[name])
+      ?.flatMap((e) => e[name])
+      ?.filter((e): e is Exclude<typeof e, undefined> => e !== undefined && e.type === type)
+      ?.map((e) => e.value)
+
+    return extensionsWithName as Array<T>
+  }
+
+  public get sanDnsNames() {
+    const subjectAlternativeNameExtensionDns = this.getMatchingExtensions<string>(id_ce_subjectAltName, 'dns')
+    return subjectAlternativeNameExtensionDns?.filter((e) => typeof e === 'string')
+  }
+
+  public get sanUriNames() {
+    const subjectAlternativeNameExtensionUri = this.getMatchingExtensions<string>(id_ce_subjectAltName, 'url')
+    return subjectAlternativeNameExtensionUri?.filter((e) => typeof e === 'string')
+  }
+
+  public static async createSelfSigned({
+    key,
+    extensions,
+    notAfter,
+    notBefore,
+    name,
+  }: {
+    key: Key
+    // For now we only support the SubjectAlternativeName as `dns` or `uri`
+    extensions?: Array<Array<{ type: 'dns' | 'uri'; value: string }>>
+    notBefore?: Date
+    notAfter?: Date
+    name?: string
+  }) {
+    const publicKey = new CryptoKey(
+      key,
+      { name: 'ECDSA', namedCurve: 'P-256', hash: { name: 'SHA-256' } },
+      true,
+      'public',
+      ['verify']
+    )
+
+    const privateKey = new CryptoKey(
+      key,
+      { name: 'ECDSA', namedCurve: 'P-256', hash: { name: 'SHA-256' } },
+      false,
+      'private',
+      ['sign']
+    )
```

Comment:
The underlying `Key` instance for the public key is the exact same as the private key as they are connected in a lower level. I think if we merge the `Key` and the `CryptoKey` more in the future it will be better, I get that this looks a bit weird, but it is.